File: Difficulty: Medium/Top View of Binary Tree/top-view-of-binary-tree.cpp
class Solution {
  public:
    vector<int> topView(Node *root) {
        // code here
        vector<pair<int,int>> temp;
        queue<pair<Node*,int>> q;
        set<int> s;
        q.push({root,0});
        while(q.empty()!=true){
            int a=q.size();
            while(a){
                pair<Node*,int> b=q.front();
                q.pop();
                if(s.find(b.second)==s.end()){
                    temp.push_back({b.second,b.first->data});
                    s.insert(b.second);
                }
                if(b.first->left!=NULL){
                    q.push({b.first->left,b.second-1});
                }
                if(b.first->right!=NULL){
                    q.push({b.first->right,b.second+1});
                }
                a--;
            }
        }
        sort(temp.begin(),temp.end());
        vector<int> ans;
        for(int i=0;i<temp.size();i++){
            ans.push_back(temp[i].second);
        }
        return ans;
    }
};

File: Difficulty: Medium/Top View of Binary Tree/top-view-of-binary-tree.java
/*
class Node {
    int data;
    Node left, right;

    Node(int val) {
        this.data = val;
        this.left = null;
        this.right = null;
    }
}
*/
class Pair {
    Node node;
    int col;
    Pair(Node node, int col) {
        this.node = node;
        this.col = col;
    }  
}

class Solution {
    public void levelOrderTraversal(Node root, TreeMap<Integer, Integer> map) {
        Deque<Pair> queue = new ArrayDeque<>();
        queue.offer(new Pair(root, 0));
        while(!queue.isEmpty()) {
            Pair pair = queue.poll();
            Node node = pair.node;
            int col = pair.col;
            map.putIfAbsent(col, node.data);
            if (node.left != null) queue.offer(new Pair(node.left, col - 1));
            if (node.right != null) queue.offer(new Pair(node.right, col + 1));
        }
    }
    public ArrayList<Integer> topView(Node root) {
        ArrayList<Integer> ans = new ArrayList<>();
        if (root == null) return ans;
        TreeMap<Integer, Integer> map = new TreeMap<>();
        levelOrderTraversal(root, map);
        for (var colEntry : map.entrySet()) {
            ans.add(colEntry.getValue());
        }
        return ans;
        
    }
}